### PR TITLE
Fluid 6250: reverting package-lock.json for dev workflow

### PR DESCRIPTION
**Pull request by atisayajain**
Fixes issue: https://issues.fluidproject.org/browse/FLUID-6250

### CODE REVIEW

**Comment by atisayajain** [comment]
@jobara I was trying to edit the commit message and removed all the commits that I made from my github fork by mistake, due to which previous PR got closed. I am really sorry for this. Will handle PR reviews better from next time.

This PR has the same changes as the previous one, you can verify and test it again.

I added `package-lock.json` where you requested and upgraded the version in `package.json`. Please let me know if I missed something somewhere, or anything else that needs to be done.

File: README.md [hunk 1]
```diff
@@ -235,10 +235,10 @@ publish.standard();
                 <code>cleanCmd</code>
             </td>
             <td>
-                The CLI to execute which cleans up any temporary changes to the package.json file.
+                The CLI to execute which cleans up any temporary changes to the package.json and package-lock.json files.
                 <ul>
                     <li>
-                        <code>${package}</code> will be substituted with the path to the executing modules package.json file.
+                        <code>${package}</code> will be substituted with the path to the executing modules package.json and package-lock.json files.
```
Comment (jobara) [inline README.md]: This line seems to have been incorrect and as such still is, because there is not substitution of a `${package}` token. Looking at the publish.clean code, it seems that the command is always executed as is, without interpolating any values. https://github.com/fluid-project/fluid-publish/blob/master/publish.js#L274-L275

Comment (atisayajain) [reply README.md]: @jobara  Yes, I also observed it while testing the command. Should I remove this line completely and also edit the corresponding default command to `git checkout -- package.json package-lock.json` in this file?

Or do you need me to make changes so that `cleanCmd` starts accepting values from command line?

Comment (jobara) [reply README.md]: @atisayajain I think the former. publish.clean will change to the specified moduleRoot first before reverting the package file. https://github.com/fluid-project/fluid-publish/blob/master/publish.js#L260-L279

So for now I think it's okay to leave the functionality as is and just update the documentation. And yes, it looks like you should remove the line and update the default command as you specified.

**Comment by atisayajain** [comment]
@jobara I have updated the documentation as per your request, please review the changes. Please let me know if anything else needs to be done.

File: package.json [hunk 2]
```diff
@@ -48,7 +48,7 @@
     "publishCmd": "npm publish",
     "publishDevCmd": "npm publish --tag ${devTag}",
     "versionCmd": "npm version --no-git-tag-version ${version}",
-    "cleanCmd": "git checkout -- package.json",
+    "cleanCmd": "git checkout -- package.json package-lock.json",
```
Comment (jobara) [inline package.json]: @atisayajain I was testing this out today and found that it isn't as robust as needed. Take for example the case where no package-lock.json exists or if it is there but is not tracked by git. In both of those cases it an error will be thrown by git. Instead we should check first if the file is being tracked by git before trying to clean it up. 

One option:
- add a new command to check if a file is being tracked by version control (see: https://stackoverflow.com/questions/2405305/how-to-tell-if-a-file-is-git-tracked-by-shell-exit-code)
- add another option that contains a list of files to clean
- update the `cleanCmd` to take a token that will be replace by each of the above files
- update `publish.clean` to iterate over the list of files to clean. It will first need to check if the file is being versioned and then call the cleanCmd on it.

**Comment by atisayajain** [comment]
@jobara Sorry for the delay. I am working on this right now. Thanks for your help, will get back soon.

Can I name the new option `cleanCmdFiles`?

**Comment by jobara** [comment]
hmm.. i'm not sure, maybe "filesToClean"

**Comment by atisayajain** [comment]
@jobara I have added two new options `checkFilesTracking` and `filesToClean`, the latter is for accepting token. Also updated `publish.clean` to check whether the files to be restored are being tracked by git or not. Please review the changes. I have also updated comments just above `publish.clean` regarding what it does, will update documentation too when you approve changes.

I have observed that most other commands have hints for them, which is displayed in case an error is thrown. Can I add one for `checkFilesTracking` too?

Please let me know if there is anything else needs to be done.

**Comment by atisayajain** [comment]
@jobara I have added a try/catch block that now iterates over each file present in `filesToClean`. Try/catch block first checks if the file is being tracked or exists, then proceeds with the cleanup. I have also added a hint because as it being displayed as `undefined` otherwise. Please review the changes.

Please let me know if there's anything else needs to be done.

File: publish.js [hunk 3]
```diff
@@ -258,22 +258,41 @@ publish.tagVC = function (isTest, version, options) {
...
  */
 publish.clean = function (moduleRoot, options) {
     var originalDir = process.cwd();
 
     // change to the module root directory
     process.chdir(moduleRoot || "./");
 
-    // run the clean command
-    publish.execSyncFromTemplate(options.cleanCmd);
+    var file = options.filesToClean.split(" ");
```
Comment (jobara) [inline publish.js]: This should probably be called "files".

File: README.md [hunk 4]
```diff
@@ -235,15 +235,10 @@ publish.standard();
...
-                <ul>
-                    <li>
-                        <code>${package}</code> will be substituted with the path to the executing modules package.json file.
-                    </li>
-                </ul>
+                The CLI to execute which cleans up any temporary changes to the package.json and package-lock.json files.
             </td>
             <td>
-                "git checkout -- ${package}"
+                "git checkout -- package.json package-lock.json"
```
Comment (jobara) [inline README.md]: Needs to be updated to the new value, and need documentation for other new options.

File: package.json [hunk 5]
```diff
@@ -48,7 +48,9 @@
     "publishCmd": "npm publish",
     "publishDevCmd": "npm publish --tag ${devTag}",
     "versionCmd": "npm version --no-git-tag-version ${version}",
-    "cleanCmd": "git checkout -- package.json",
+    "checkFilesTracking": "git ls-files --error-unmatch ${filesToClean}",
```
Comment (jobara) [inline package.json]: this only deals with a single file, so you probably want to take in the file and not filesToClean. Also the name should reflect that it only checks a single file at a time.

File: publish.js (hunk 3 above)
Comment (jobara) [reply publish.js]: Also i think it would make more sense to expect a comma separated list.

File: publish.js [hunk 6]
```diff
@@ -258,22 +258,41 @@ publish.tagVC = function (isTest, version, options) {
...
+    
+    file.forEach(function(file){
+        var isTracked = true;
 
+        // check whether file is being tracked or not
+        try {publish.execSyncFromTemplate(options.checkFilesTracking, {
+                filesToClean: file
+            }, options.checkFilesTrackingHint);
+        } catch (error) {
```
Comment (jobara) [inline publish.js]: is there a way to check for a specific error, just in case something else goes wrong, we may not want to treat all errors as the file not being tracked.

File: publish.js [hunk 7]
```diff
@@ -258,22 +258,41 @@ publish.tagVC = function (isTest, version, options) {
...
+            isTracked = false;
+        }
+        
+        if (isTracked)
+            // run the clean command if file is being tracked
+            publish.execSyncFromTemplate(options.cleanCmd, {
+                filesToClean: file
+            });
```
Comment (jobara) [inline publish.js]: You might be able to simplify this by putting this step in the try block as well. Assuming you can differentiate between types of errors.

File: package.json [hunk 8]
```diff
@@ -59,6 +61,7 @@
     "checkRemoteHint": "Run \"git remote -v\" for a list of available remote repositories.\n",
     "publishHint": "Ensure that you have access to publish to the registry and that the current version does not already exist.\n",
     "publishDevHint": "Ensure that you have access to publish to the registry and that the current version does not already exist.\nIf the npm tag specified by --tag is recognizable as a valid semver version number, it will be rejected by npm. This is because version numbers and tags share a common namespace for npm packages.\n",
+    "checkFilesTrackingHint": "Either the file ${filesToClean} is not being tracked or does not exist.",
```
Comment (jobara) [inline package.json]: Again i think this actually refers to a single file as opposed to many, and should be adjusted to reflect that. Also maybe reword to something like "${file} was not cleaned. It is either not being tracked or does not exist."

**Comment by jobara** [comment]
@atisayajain could you also add any necessary unit tests for the changes you are making.

File: publish.js (hunk 6 above)
Comment (atisayajain) [reply publish.js]: Hi, @jobara. It is that git returns the full name of the file if it is present and is being tracked. Can I use this as a parameter to differentiate between errors?

Comment (jobara) [reply publish.js]: hmm.. i'm not sure that will help distinguish the errors, because if an occurs for any reason, you may not get the file name back.

Comment (atisayajain) [reply publish.js]: @jobara I am now able to extract return codes for errors that I think is a better approach. Here, git returns 0 if command was run successfully and 1 if the file hasn't been added or is being tracked. What should be the expected behaviour for error codes other than 0 and 1?

Comment (jobara) [reply publish.js]: That would be the exit code, which might be useful depending on what other exit codes are produced. However, it's possible that there may be multiple reasons for exiting with the same code when in an error state. Is there any useful information provided in the error object that you get in the catch?

File: package.json [hunk 9]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "fluid-publish",
-  "version": "2.0.0",
+  "version": "2.0.1",
```
Comment (jobara) [inline package.json]: You'll need to update from master as a new change has already gone in and has bumped the version to 2.1.0. You'll probably want to set your version to 2.1.1.

File: publish.js (hunk 6 above)
Comment (atisayajain) [reply publish.js]: @jobara Error object also returns an error message in form of: 
```
Command failed: git ls-files --error-unmatch package-lock.json
error: pathspec 'package-lock.json' did not match any file(s) known to git.
Did you forget to 'git add'?
```
How do you suggest using this information?

File: package.json (hunk 9 above)
Comment (atisayajain) [reply package.json]: Sure, I will.

File: publish.js (hunk 6 above)
Comment (jobara) [reply publish.js]: It doesn't sound like we get a lot of detail on the exact error except through the message returned. I think we will want to log a message that indicates that we suspect that that the file is not tracked or not present and provide the error message that was provided. This way if something else is failing at least the user will know what happened.

**Comment by atisayajain** [comment]
@jobara Please review the changes and let me know if there's anything else that needs to be done.

Will add the unit tests too, after your review.

File: README.md [hunk 10]
```diff
@@ -241,20 +241,47 @@ publish.standard();
                 </p>
             </td>
         </tr>
+        <tr>
+            <td>
+                <code>checkFileTracking</code>
+            </td>
+            <td>
+                The CLI to execute which checks if there are any temporary changes.
```
Comment (jobara) [inline README.md]: Probably want to mention that these were changes made as part of the publishing, for example the version number in the package.json file.

File: README.md [hunk 11]
```diff
@@ -241,20 +241,47 @@ publish.standard();
...
+                "git ls-files --error-unmatch ${file}"
+            </td>
+        </tr>
         <tr>
             <td>
                 <code>cleanCmd</code>
             </td>
             <td>
-                The CLI to execute which cleans up any temporary changes to the package.json file.
+                The CLI to execute which cleans up any temporary changes.
```
Comment (jobara) [inline README.md]: Similar to above comment for "checkFileTracking"

File: README.md [hunk 12]
```diff
@@ -241,20 +241,47 @@ publish.standard();
...
-                "git checkout -- ${package}"
+                "git checkout -- ${file}"
+            </td>
+        </tr>
+        <tr>
+            <td>
+                <code>filesToClean</code>
+            </td>
+            <td>
+                The names of files to be checked and cleaned after the test release.
```
Comment (jobara) [inline README.md]: Probably good to specify that it is a comma separated list of file names.

**Comment by jobara** [comment]
@atisayajain Thanks for the update. I've left a few more comments, but in general I think it's ready for you to start writing some tests. Also please make sure to run `grunt lint` from the command line in the project root. I've noticed that there are some code linting errors.

**Comment by atisayajain** [comment]
@jobara This might be silly but I need some help in writing tests here. Could you please tell me more about this particular  error `AssertionError [ERR_ASSERTION]: execSync should have been called`?

**Comment by the-t-in-rtf** [comment]
Hi, @atisayajain.  It would  be easier if we could see the error in context.  I'm happy to take a look if you can commit your work in progress to the branch associated with this pull.

File: tests/publishTests.js [hunk 13]
```diff
@@ -431,7 +431,9 @@ console.log("\n*** publish.clean ***");
 
 var cleanFixture = [{
     moduleRoot: "./",
-    cleanCmd: "clean"
+    checkFileTracking: "checkFileTracking",
+    cleanCmd: "clean",
+    filesToClean: "package.json, package-lock.json"
 }];
```
Comment (atisayajain) [inline tests/publishTests.js]: @the-t-in-rtf I have added just the options for the changes I have made. On running `npm test`, I am getting an assertion error in publish.clean

**Comment by the-t-in-rtf** [comment]
Hi, @atisayajain.  Before we continue, I need to point out that we had a discussion in our team meeting last night, and for unrelated reasons we have agreed to stop committing package-lock.json files.  So, this pull is no longer needed and will not be merged.

However, out of respect for the time you put in, and to help with your learning process, I took a few minutes to look at your branch and figured out why it wasn't running.  In short, there were a few drawbacks, the largest of which are:

1. You did not update the failing check to reflect the actual number of times the execSync function was called (you had two calls per file, and two files, so you had four calls instead of the previous one).
2. You did not update the test definitions to match the defaults you added to the package.json file (you would have seen this problem as soon as you fixed the previous issue).

I will leave additional feedback in-context so that you can review for your own learning process.  Thanks again for continuing to work on this, even if the team no longer needs this improvement.  I hope we can find another opportunity to work together in the future.

File: package.json [hunk 14]
```diff
@@ -49,7 +49,9 @@
     "publishCmd": "npm publish",
     "publishDevCmd": "npm publish --tag ${devTag}",
     "versionCmd": "npm version --no-git-tag-version ${version}",
-    "cleanCmd": "git checkout -- package.json",
+    "checkFileTracking": "git ls-files --error-unmatch ${file}",
```
Comment (the-t-in-rtf) [inline package.json]: You should have added all of these new defaults to the publishFixture definitions in the tests.  That was the cause of one failure.

File: publish.js [hunk 15]
```diff
@@ -266,21 +266,37 @@ publish.tagVC = function (isTest, version, options) {
...
     var originalDir = process.cwd();
 
     // change to the module root directory
     process.chdir(moduleRoot || "./");
 
-    // run the clean command
-    publish.execSyncFromTemplate(options.cleanCmd);
+    var file = options.filesToClean.split(/\s*,\s*/);
+
+    file.forEach(function (file) {
```
Comment (the-t-in-rtf) [inline publish.js]: For future reference, it is confusing and dangerous to reuse variable names in the way you've done here.  It would have been better to name the array `files`, for example.

File: publish.js [hunk 16]
```diff
@@ -266,21 +266,37 @@ publish.tagVC = function (isTest, version, options) {
...
+
+        // check whether file is being tracked or not and then try to clean it
+        try {
+            publish.execSyncFromTemplate(options.checkFileTracking, {
```
Comment (the-t-in-rtf) [inline publish.js]: I understand what you were trying to do here, but this new check does nothing.  There is no return value from the call and you don't actually use that to decide whether to clean the file.

Comment (the-t-in-rtf) [reply publish.js]: Or perhaps you were counting on an error to be thrown if the file was untracked?  If that's the case, a comment would be a good idea.

Comment (the-t-in-rtf) [reply publish.js]: OK, I had to learn a bit more about how this package works myself (that's normal).  I see that it would have thrown an error if the file wasn't tracked, that part would be more or less OK with a reasonable comment.

**Comment by the-t-in-rtf** [comment]
So, in a final comment, I wanted to share a bit about how I actually diagnosed the underlying problem.  For starters, as I had no previous experience with Sinon, I read their [API docs](http://sinonjs.org/releases/v1.17.7/) to find out what the failing call did, and what other calls were available.

I then ran the tests in a debugger (I use one built into my IDE) with a breakpoint set right before the failing test is called (there was a line number in the error).  I then inspected the Sinon "stub" to see how it differed from the condition being tested.  I could see that `calledOnce` returns true if the function was called once, and looking at the stub's `callCount` property, I could see that the function had been called four times.  Looking at your code, I could see this was to be expected, and replaced `stub.execSync.calledOnce` with `stub.execSync.callCount === 4`.

Once I did that, the second failure appeared [here](https://github.com/atisayajain/fluid-publish/blob/3c177b4dd4961f353071500ee63e5588fee7a6c9/tests/publishTests.js#L566).  I was able to diagnose this by comparing `fixture.options` to the output of `stub.execSync.getCalls()`.  It became clear that `fixture.options` simply lacked the new default values you had added.  As this test framework exits on the first failure, it only appeared to be one problem, but really, there would have been quite a few other failures of this type.

In both cases, the process of figuring out the issue was roughly:

1. Find the location of the failing test.
2. Pause the tests before the failure.
3. Look at what is  being compared, the expected value, and the actual value.

I hope that's helpful to you, and thanks again for the time you put in.

**Comment by atisayajain** [comment]
Hi @the-t-in-rtf, it is alright that this branch won't be merged, I hope to find something soon enough on JIRA for me.

And thank you so much for putting in effort and clearing my doubts regarding the unit tests. Honestly, I got to learn a lot just in the course of this PR. This means a lot to me from a learning point of view. Thanks, again.

**Comment by amb26** [comment]
Thanks for your work on this, @atisayajain - as @the-t-in-rtf mentioned, in the end we have decided to avoid committing package-lock.json files to project repositories so this functionality won't be used. Thanks also to @the-t-in-rtf for the time and care he put in for review and diagnosing problems.
